Testscripts/Linux/perf_apache.sh

```diff
@@ -0,0 +1,255 @@
+
+Run_ab()
+{
+	current_concurrency=$1
+
+	if [ ${current_concurrency} -le 2 ]
+	then
+		total_requests=50000
+	elif [ ${current_concurrency} -le 128 ]
+	then
+		total_requests=100000
+	else
+		total_requests=200000
+	fi
+
+	ab_instances=$(($current_concurrency / $max_concurrency_per_ab))
+	if [ ${ab_instances} -eq 0 ]
+	then
+		ab_instances=1
+	fi
+	if [ ${ab_instances} -gt ${max_ab_instances} ]
+	then
+		ab_instances=${max_ab_instances}
+	fi
+
+	total_request_per_ab=$(($total_requests / $ab_instances))
+	concurrency_per_ab=$(($current_concurrency / $ab_instances))
+	concurrency_left=${current_concurrency}
+	requests_left=${total_requests}
+	while [ ${concurrency_left} -gt ${max_concurrency_per_ab} ]; do
+		concurrency_left=$(($concurrency_left - $concurrency_per_ab))
+		requests_left=$(($requests_left - $total_request_per_ab))
+		LogMsg "Running parallel ab command for: ${total_request_per_ab} X ${concurrency_per_ab}"
+		ab -n ${total_request_per_ab} -r -c ${concurrency_per_ab} http://${server}/test.dat >> ${log_folder}/${current_concurrency}.${testName}.bench.log & pid=$!
+		PID_LIST+=" $pid"
+	done
+
+	if [ ${concurrency_left} -gt 0 ]
+	then
+		LogMsg "Running parallel ab command left for: ${requests_left} X ${concurrency_left}"
+		ab -n ${requests_left} -r -c ${concurrency_left} http://${server}/test.dat >> ${log_folder}/${current_concurrency}.${testName}.bench.log & pid=$!
+		PID_LIST+=" $pid";
+	fi
+	trap "kill ${PID_LIST}" SIGINT
+	wait ${PID_LIST}
+}
+
+Run_apache()
+{
+	Run_SSHCommand "${server}" "pkill -f ab"
+	pkill -f ab
+
+	# Restart apache2 service
+	LogMsg "Info: Generate test data file on the Apache server /var/www/html/test.dat"
+	Run_SSHCommand "${server}" "dd if=/dev/urandom of=${document_path} bs=1K count=200"
+	Run_SSHCommand "${server}" "service ${web_server} stop"
+	Run_SSHCommand "${server}" "service ${web_server} start"
+	
+	Run_SSHCommand "${server}" "mkdir -p ${log_folder}"
+	mkdir -p ${log_folder}
+	result_file="${log_folder}/report.csv"
+	echo "WebServerVersion,TestConcurrency,NumberOfAbInstances,ConcurrencyPerAbInstance,Document_bytes,CompleteRequests,RequestsPerSec,TransferRate_KBps,MeanConnectionTimes_ms" > "${result_file}"
+
+	for current_concurrency in "${testConcurrency[@]}"; do
+		LogMsg "Running apache_bench test with current concurrency: ${current_concurrency}"
+		Run_SSHCommand "${server}" "pkill -f dstat"
+		Run_SSHCommand "${server}" "pkill -f mpstat"
+		Run_SSHCommand "${server}" "pkill -f sar"
+		Run_SSHCommand "${server}" "pkill -f vmstat"
+		Run_SSHCommand "${server}" "${dstat_cmd} -dam" > "${log_folder}/${current_concurrency}.dstat.server.log" &
+		Run_SSHCommand "${server}" "${sar_cmd} -n DEV 1" > "${log_folder}/${current_concurrency}.sar.server.log" &
+		Run_SSHCommand "${server}" "${mpstat_cmd} -P ALL 1" > "${log_folder}/${current_concurrency}.mpstat.server.log" &
+		Run_SSHCommand "${server}" "${vmstat_cmd} 1" > "${log_folder}/${current_concurrency}.vmstat.server.log" &
+
+		pkill -f dstat
```

Comment:
Why are we killing this processes, is it not existing gracefully ?

Reply:
If we don't kill this process, "dstat" won't exit.

Reply:
Man page of dstat show below options. Will count help here to exit after fixed time period ?
Usage: dstat [-afv] [options..] [delay ```[count]```]

Reply:
Thank you for telling me this. In this apache test, we don't know the accurately time that the ab test will cost.